File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int dp[13][10001];
    
    int coinChange(vector<int>& coins, int amount) {
        memset(dp, -1, sizeof dp);
        int ans = min(findWays(coins, 0, amount), findWays(coins, 1, amount));
        if(ans == INT_MAX - 1) return -1;
        return ans;
    }
    
    int findWays(vector<int>& coins, int index, int amount) {
        if(amount < 0 || index >= coins.size()) return INT_MAX - 1;
        if(dp[index][amount] != -1) return dp[index][amount];
        if(amount == 0) return 0;
        return dp[index][amount] = min(1 + findWays(coins, index, amount - coins[index]), findWays(coins, index + 1, amount));
    }
};

File: 0322-coin-change/0322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function (coins, amount) {
  const dp = Array.from({ length: 13 }, () => []);

  const calculateCoin = (index, amount) => {
    if (amount == 0) return 0;
    if (dp[index][amount] !== undefined) return dp[index][amount];
    if (amount < 0 || index >= coins.length) return Number.POSITIVE_INFINITY;

    dp[index][amount] = Math.min(1 + calculateCoin(index, amount - coins[index]), calculateCoin(index + 1, amount));
    return dp[index][amount];
  }

  const ans = calculateCoin(0, amount);
  return ans === Number.POSITIVE_INFINITY ? -1 : ans;

};